---
title: "R Notebook"
# output: html_notebook
---



Load the libraries
```{r}

# rm(list=ls())

library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(tidyverse)
library(lubridate)
library(RColorBrewer)
library(stringr)
library(ComplexHeatmap)
library(varhandle)
library(ggpubr)
library(Seurat)
library(cowplot)
library(dplyr)
library(hash)

options(ggrepel.max.overlaps = Inf)

set.seed(777)

```


Load custome-written functions
```{r}

dir<-"Rutilities_v0.1"

files = list.files(path=dir, full.names = TRUE)

files

sapply(files, source)


dir<-"IntegrationUsingOrthogroup_v0.1"

files = list.files(path=dir, pattern="*.R",  full.names = TRUE)

files

sapply(files, source)



dir<-"StoufferDEG_v0.1"

files = list.files(path=dir, pattern="*.R",  full.names = TRUE)

files

sapply(files, source)

dir<-"TranscriptomeMapping_v0.4"

files = list.files(path=dir, pattern="*.R",  full.names = TRUE)

files

sapply(files, source)



```



Load the lamprey dataset
```{r}

load("seu.lamprey.RGC.types.rda")

Idents(seu)<-"orig.ident"

VlnPlot(seu, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

seu@meta.data$orig.ident %>% table()


seu@meta.data$sample<-seu@meta.data$orig.ident

seu@meta.data$subtype %>% unique() %>% sort()


seu@meta.data$subtype<-paste0("RGC", seu@meta.data$subtype)

p1<-DimPlot(seu, reduction = "umap",
            group.by="subtype",
            label = TRUE,
            repel=TRUE,
            cols=col_vector,
            raster=FALSE,
            label.box=FALSE)
p1
p1<-p1+gg.theme+labs(title="")
p1


seu@meta.data$subtype %>% table()


Idents(seu)<-"subtype"
seu<-subset(seu, downsample=200)
seu@meta.data$subtype %>% table()
Idents(seu)<-"subtype"

#

DefaultAssay(seu)<-"RNA"

seu@meta.data$orig.ident<-seu@meta.data$orig.ident %>% as.vector() %>% as.factor()

seu@meta.data$species<-"lamprey"

seu@meta.data$cell_type<-seu@meta.data$subtype

seu@meta.data$subtype2<-seu@meta.data$subtype

seu.lamprey<-seu

rm(seu)


```




Load the zebrafish dataset
```{r}


seu<-readRDS("adult_zFish_FINAL.rds")

seu

seu@meta.data$cell_type<-seu@meta.data$clusterID

seu@meta.data$clusters<-seu@meta.data$clusterID

seu@meta.data$subtype<-seu@meta.data$clusterID

seu@meta.data$species<-"zebrafish"

counts<-seu@assays$RNA@counts

tmp<-unique(rownames(counts))

counts<-counts[tmp,]

seu<-CreateSeuratObject(counts=counts, meta.data=seu@meta.data)



Idents(seu)<-"orig.ident"

VlnPlot(seu, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

seu@meta.data$orig.ident %>% table() %>% sort(decreasing = TRUE)

tmp<-seu@meta.data$orig.ident %>% table() %>% sort(decreasing = TRUE) %>% names()

tmp

seu<-subset(seu, idents=tmp[1:3])

#
Idents(seu)<-"subtype"

seu<-subset(seu, downsample=200)

seu@meta.data$subtype %>% table() %>% sort(decreasing = TRUE)

seu.zebrafish<-seu 


```






```{r}

seu<-seu.lamprey

Idents(seu)<-"orig.ident"

samples<-seu@meta.data$orig.ident %>% table() %>% sort(decreasing = TRUE) %>% names()

seu<-subset(seu, idents=samples[1])

seu<-seu%>% NormalizeData() %>% FindVariableFeatures() 

seu<-ScaleData(seu, features =rownames(seu))

seu[['RNA']]<-as(seu[['RNA']], Class="Assay")

seu.lamprey.1<-seu

```




```{r}

seu<-seu.zebrafish
  
Idents(seu)<-"orig.ident"

samples<-seu@meta.data$orig.ident %>% table() %>% sort(decreasing = TRUE) %>% names()

seu<-subset(seu, idents=samples[1])

DefaultAssay(seu)<-"RNA"

seu<-seu%>% NormalizeData() %>% FindVariableFeatures() 

seu<-ScaleData(seu, features =rownames(seu))

seu[['RNA']]<-as(seu[['RNA']], Class="Assay")

seu.zebrafish.1<-seu
```


Assay converting
```{r}
seu.lamprey[["RNA"]]<-as(object=seu.lamprey[["RNA"]], Class="Assay")
seu.zebrafish[["RNA"]]<-as(object=seu.zebrafish[["RNA"]], Class="Assay")

```



Edit orthgroup table
```{r}

genes_zebrafish<-rownames(seu.zebrafish)
genes_lamprey<-rownames(seu.lamprey)


table<-read.table(file="Zebrafish__v__Lamprey.tsv",
                  fill=TRUE,
                  sep="\t",
                  header=TRUE)

# Filtering the genes in the table not present in the expression matrix
tmp<-FilterOrthologuesByExpression(x=table$Zebrafish,
                                   y=genes_zebrafish)

table$Zebrafish<-tmp

# 
tmp<-FilterOrthologuesByExpression(x=table$Lamprey,
                                   y=genes_lamprey)

table$Lamprey<-tmp


# remove the rows including the NA values
table<-table %>%
  as.data.frame() %>%
  dplyr::filter(Zebrafish!="NA") %>%
  dplyr::filter(Lamprey!="NA")

table<-apply(table, MARGIN=c(1,2), FUN=MakeOrthoUnique)

table <-  FilterOrthoByOrthoSize(x=table, size_max = 5)

#

# Select the best matched 
table<-GetBestOrthologuesBySD(ortho_group_table=table,
                               exp_mat=seu.lamprey.1@assays$RNA@scale.data,
                              species="Lamprey",
                              pattern=","
                              )


table<-GetBestOrthologuesBySD( ortho_group_table=table,
                              exp_mat=seu.zebrafish.1@assays$RNA@scale.data,
                              species="Zebrafish",
                              pattern=","
                              )


dim(table)



# remove the rows including the NA values
table<-table %>%
  as.data.frame() %>%
  dplyr::filter(Zebrafish!="NA") %>%
  dplyr::filter(Lamprey!="NA")

table<-apply(table, MARGIN=c(1,2), FUN=MakeOrthoUnique)

table <-  FilterOrthoByOrthoSize(x=table, size_max = 5)

# Remove duplicated lines
table<-unique(table,  MARGIN=1)

message("orthogroup numbers in the processed table are:", nrow(table))

rownames(table)<-table$Orthogroup %>% make.unique()

ortho_group_table_reference<-table

save(ortho_group_table_reference, file="ortho_group_table_reference.rda")




```


Aggregate expressions for orthogroups
```{r}

mat<-AggregateOrthoGroup(exp_mat=seu.zebrafish@assays$RNA@counts,
                              ortho_group_table=ortho_group_table_reference,
                              column_species_index=2,
                              pattern=","
)


mat<-OrthogroupID2GeneSymbol(mat=mat,
                                  orthogroup_table=ortho_group_table_reference,
                                  reference_species=c("Lamprey", "Zebrafish")
                                  )

dim(mat)

mat[1:3,1:3]

df<-seu.zebrafish@meta.data

meta_data<-df[, -which(colnames(df) %in% c("nCount_RNA", "nFeature_RNA"))]

seu.zebrafish<-CreateSeuratObject(counts=mat, meta.data =meta_data)

seu.zebrafish@meta.data$cell_type %>% table()

#------------------------------------
# for the lamprey

mat<-AggregateOrthoGroup(exp_mat=seu.lamprey@assays$RNA@counts,
                              ortho_group_table=ortho_group_table_reference,
                              column_species_index=3,
                              pattern=","
)


mat<-OrthogroupID2GeneSymbol(mat=mat,
                                  orthogroup_table=ortho_group_table_reference,
                                  reference_species=c("Lamprey", "Zebrafish")
                                  )

dim(mat)

mat[1:3,1:3]

dim(mat)

df<-seu.lamprey@meta.data

meta_data<-df[, -which(colnames(df) %in% c("nCount_RNA", "nFeature_RNA"))]

seu.lamprey<-CreateSeuratObject(counts=mat, meta.data = meta_data)

seu.lamprey@meta.data$cell_type %>% table()


```




Merge the datasets 
```{r}

seu<-merge(x=seu.zebrafish, y=seu.lamprey)

tmp<-seu.zebrafish@meta.data$orig.ident %>% unique() %>% length()


```


Do integration
```{r}

#
seu@meta.data$species<- seu@meta.data$species %>% as.factor()
seu@meta.data$orig.ident<- seu@meta.data$orig.ident %>% as.factor()

#
Idents(seu)<-"orig.ident"

DefaultAssay(seu)<-"RNA"

seu.list<-SplitObject(seu, split.by = "orig.ident")

seu.list

seu.list <- lapply(X = seu.list, 
                   FUN = SCTransform, 
                   method = "glmGamPoi", 
                   vst.flavor = "v1",
                   conserve.memory=FALSE, 
                   return.only.var.genes=TRUE)


features1<-SelectIntegrationFeatures(object.list = seu.list[c(1:tmp)], nfeatures = 3000)
features2<-SelectIntegrationFeatures(object.list = seu.list[-c(1:tmp)], nfeatures = 3000)
features<-intersect(features1, features2)

# features <- SelectIntegrationFeatures(object.list = seu.list, nfeatures = 3000)

seu.list <- PrepSCTIntegration(object.list = seu.list, anchor.features = features)

seu.list <- lapply(X = seu.list, FUN = function(x) {
  x <- RunPCA(x, features = features, verbose = FALSE)
})


```



Do integration 
```{r}

#
anchors <- FindIntegrationAnchors(object.list = seu.list, 
                                  # reference=ref.id,  
                                  normalization.method = "SCT", 
                                  scale=TRUE,
                                  anchor.features = features,
                                  reduction = "rpca",
                                  k.anchor = 20,
                                  dims = 1:50)

# rm(seu.list)


```

```{r}

seu <- IntegrateData(anchorset = anchors, 
                     normalization.method = "SCT", 
                  #  k.weight = 50,
                     dims = 1:50)

# rm(anchors)

```




```{r}
#DefaultAssay(seu)<-"integrated"

seu <- RunPCA(seu, verbose = FALSE)
seu <- RunUMAP(seu, reduction = "pca", dims = 1:50)
seu <- FindNeighbors(seu, reduction = "pca", dims = 1:50)
seu <- FindClusters(seu, resolution = 0.8)

seu@meta.data$seurat_clusters<-seu@meta.data$seurat_clusters %>% as.numeric() %>% as.factor()

seu@meta.data$integrated_clusters<-seu@meta.data$seurat_clusters
```


Plot the integration
```{r}

dir_out<-"figures_integration"

if(! dir.exists(dir_out)){
  
  dir.create(dir_out)
}

setwd(dir_out)


# Do plot 
# cols2<-c("grey", "red")

 cols2<-c("grey", "blue") %>% rev()

# cols2<-c("#B5CA16", "#EE6C84")

p1<-DimPlot(seu, reduction = "umap", 
            group.by = "species", 
          #  order=rev(rownames(seu@meta.data)),
            label = FALSE, 
            cols=cols2,
            repel=TRUE)

p1

p1<-p1+gg.theme3

p1

file<- paste0("Dimplot_species", ".pdf")

ggsave(file=file, width=5.5, height=5)


#--
# cols2<-c("grey", "blue")

p1<-DimPlot(seu, reduction = "umap", 
            group.by = "seurat_clusters", 
          #  order=rev(rownames(seu@meta.data)),
            label = TRUE, 
            repel=TRUE)

p1

p1<-p1+gg.theme3

p1

file<- paste0("Dimplot_seurat_clusters", ".pdf")

ggsave(file=file, width=5.5, height=5)




#------

p1<-DimPlot(seu, reduction = "umap", 
            group.by = "species", 
           # order=rev(rownames(seu@meta.data)),
            label = FALSE, 
            cols=cols2,
            repel=TRUE)

p1

p1<-p1+gg.theme3+NoLegend()

p1

p11<-p1

file<- paste0("Dimplot_species_Nolgend_width5.5_height5", ".pdf")

ggsave(file=file, width=5.5, height=5)



#------subtype 

Idents(seu)<-"species"

seu.zebrafish<-subset(seu, idents="zebrafish")

seu.lamprey<-subset(seu, idents="lamprey")


p3<-DimPlot(seu.zebrafish, reduction = "umap", 
            group.by="subtype", 
            cols=col_vector,
            label = TRUE, repel=TRUE)

p3

p3<-p3+gg.theme3

p3


file<- paste0("Dimplot_subtypes_zebrafish", ".pdf")

ggsave(file=file, width=6, height=5)


#------

p3<-DimPlot(seu.zebrafish, reduction = "umap", 
            group.by="subtype", 
            cols=col_vector,
            label = TRUE, repel=TRUE)

p3

p3<-p3+gg.theme3+NoLegend()

p3

p22<-p3

file<- paste0("Dimplot_subtypes_zebrafish_Nolgend_width5.5_height5", ".pdf")

ggsave(file=file, width=5.5, height=5)



#-------------zebrafish

p3<-DimPlot(seu.lamprey, reduction = "umap", 
            group.by="subtype", 
            cols=col_vector,
            label = TRUE, repel=TRUE)

p3

p3<-p3+gg.theme3

p3

file<- paste0("Dimplot_subtypes_lamprey", ".pdf")

ggsave(file=file, width=6, height=5)


#-----------

p3<-DimPlot(seu.lamprey, reduction = "umap", 
            group.by="subtype", 
            cols=col_vector,
            label = TRUE, 
            repel=TRUE)

p3

p3<-p3+gg.theme3+NoLegend()

p3

p33<-p3

file<- paste0("Dimplot_subtypes_lamprey_Nolgend_width5.5_height5", ".pdf")

ggsave(file=file, width=5.5, height=5)


#--------split by species

p4<-DimPlot(seu, reduction = "umap", group.by="subtype", label = TRUE, 
            repel=TRUE, split.by = "species",
            ncol=2)+NoLegend()

p4

file<- paste0("Dimplot_subtypes_splitbySpeceis", ".pdf")

ggsave(file=file, width=5, height=10)


```


plot the combined figures
```{r}

dir_out<-"figures_integration"

if(! dir.exists(dir_out)){
  
  dir.create(dir_out)
}

setwd(dir_out)

wrap_plots(list(p11, p33, p22), ncol = 3)

ggsave("dimplot.integrated.pdf", width=15, height=5)


```





